examples/libs/bitcoinjs/taproot/main.ts

```diff
@@ -0,0 +1,266 @@
+
+  const p2pk_psbt = new Psbt({ network });
+  p2pk_psbt.addInput({
+    hash: utxos[0].txid,
+    index: utxos[0].vout,
+    witnessUtxo: { value: utxos[0].value, script: p2pk_p2tr.output! },
+    tapLeafScript: [
+      {
+        leafVersion: p2pk_redeem.redeemVersion,
+        script: p2pk_redeem.output,
+        controlBlock: p2pk_p2tr.witness![p2pk_p2tr.witness!.length - 1]
+      }
+    ]
+  });
+
+  p2pk_psbt.addOutput({
+    address: "tb1plaw5a9rpyezadkma95ghqmwjcmp9w0t8fyvgfx52m3hn2pqu0t8qwtzhpp", // faucet address
+    value: utxos[0].value - 200
+  });
+
+  await p2pk_psbt.signInputAsync(0, wallet);
+  p2pk_psbt.finalizeAllInputs();
+
+  tx = p2pk_psbt.extractTransaction();
+  console.log(`Broadcasting Transaction Hex: ${tx.toHex()}`);
+  txid = await broadcast(tx.toHex());
+  console.log(`Success! Txid is ${txid}`);
+
+
+  console.log(`Waiting till UTXO is detected at this Address: ${script_addr}`);
+  utxos = await waitUntilUTXO(script_addr)
+  console.log(`Trying the Hash lock spend path with UTXO ${utxos[0].txid}:${utxos[0].vout}`);
+
+  const tapLeafScript = {
+    leafVersion: hash_lock_redeem.redeemVersion,
+    script: hash_lock_redeem.output,
+    controlBlock: hash_lock_p2tr.witness![hash_lock_p2tr.witness!.length - 1]
+  };
+
+  const psbt = new Psbt({ network });
+  psbt.addInput({
+    hash: utxos[0].txid,
+    index: utxos[0].vout,
+    witnessUtxo: { value: utxos[0].value, script: hash_lock_p2tr.output! },
+    tapLeafScript: [
+      tapLeafScript
+    ]
+  });
+
+  psbt.addOutput({
+    address: "tb1plaw5a9rpyezadkma95ghqmwjcmp9w0t8fyvgfx52m3hn2pqu0t8qwtzhpp", // faucet address
+    value: utxos[0].value - 200
+  });
+
+  psbt.signInput(0, hash_lock_keypair);
+
+  // We have to construct our witness script in a custom finalizer
+  const customFinalizer = (_inputIndex: number, input: any) => {
+    const scriptSolution = [
+      input.tapScriptSig[0].signature,
+      secret_bytes
+    ];
+    const witness = scriptSolution
+      .concat(tapLeafScript.script)
+      .concat(tapLeafScript.controlBlock);
+
+    return {
+      finalScriptWitness: witnessStackToScriptWitness(witness)
+    }
+  }
+
+  psbt.finalizeInput(0, customFinalizer);
+
+  tx = psbt.extractTransaction();
+  console.log(`Broadcasting Transaction Hex: ${tx.toHex()}`);
+  txid = await broadcast(tx.toHex());
+  console.log(`Success! Txid is ${txid}`);
+
+  // We can also spend from this address without using the script tree
+  console.log(`Waiting till UTXO is detected at this Address: ${script_addr}`);
+  utxos = await waitUntilUTXO(script_addr)
+  console.log(`Trying the P2PK spend path with UTXO ${utxos[0].txid}:${utxos[0].vout}`);
+
+  const key_spend_psbt = new Psbt({ network });
+  key_spend_psbt.addInput({
+    hash: utxos[0].txid,
+    index: utxos[0].vout,
+    witnessUtxo: { value: utxos[0].value, script: script_p2tr.output! },
+    tapInternalKey: toXOnly(wallet.rootPublicKey),
+    tapMerkleRoot: script_p2tr.hash
```

Comment:
so we place the `taprootMerkleRoot` in the psbt input, doesn't this mean we should be able to ship the whole psbt server side to sign?

Reply:
Yes I think we could extract it, however we won't be able to do just:

```
 const psbt = Psbt.fromBuffer(hexToBuffer(unsignedData))
 await psbt.signAllInputsAsync(signer, [Transaction.SIGHASH_ALL])
```

because in the `SignerAsync` callbacks, `signSchnorr?(hash: Buffer)` does not have any hint on what is the `taprootMerkleRoot`

we would have to iterate through the inputs and sign one by one.
I guess it is doable, do you think we should update the bik code?